compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -8701,23 +8702,15 @@ private BType checkTupleIndexBasedAccess(BLangIndexBasedAccess accessExpr, BTupl
                         }
                     }
                 });
-
-                BFiniteType finiteType;
-                if (finiteTypes.size() == 1) {
-                    finiteType = finiteTypes.get(0);
-                } else {
-                    Set<BLangExpression> valueSpace = new LinkedHashSet<>();
-                    finiteTypes.forEach(constituent -> valueSpace.addAll(constituent.getValueSpace()));
-                    finiteType = new BFiniteType(null, valueSpace);
-                }
-
-                BType possibleType = checkTupleIndexBasedAccess(accessExpr, tuple, finiteType);
-                if (possibleType.tag == TypeTags.UNION) {
-                    possibleTypesByMember.addAll(((BUnionType) possibleType).getMemberTypes());
-                } else {
-                    possibleTypesByMember.add(possibleType);
+                if (!finiteTypes.isEmpty()) {
+                    BFiniteType finiteType = createFiniteTypeFromFiniteTypeList(finiteTypes);
+                    BType possibleType = checkTupleIndexBasedAccess(accessExpr, tuple, finiteType);
+                    if (possibleType.tag == TypeTags.UNION) {
+                        possibleTypesByMember.addAll(((BUnionType) possibleType).getMemberTypes());
```

Comment:
Do we have a test that reach this point?

Reply:
try this.
```
public function main() {
    [int, string] tup = [1, "x"];
    0|1|int a = 0;
    tup[a] = 4;
}
```

Reply:
Added suggested example.